File: Linked List/linked.cpp
#include <iostream>
#include <vector>
using namespace std;
class Node{
  public:
    int data;
    Node* next;
public:
Node(int data1, Node* next1){
    data = data1;
    next = next1;
}
Node(int data1){
    data = data1;
    next = nullptr;
}
};
Node* convertArrToLL(vector<int>& arr){
    Node* head = new Node(arr[0]);
    Node* mover = head;
    for(int i = 1; i <= arr.size() - 1; i++){
        Node* temp = new Node(arr[i]);
        mover->next = temp;
        mover = temp;
    }
    return head;
}
void traverseLL(Node* head){
    Node* temp = head;
    while(temp){
        cout << temp->data <<endl;
        temp = temp->next;
    }
}
int countLength(Node* head){
    Node* temp = head;
    int count = 0;
    while(temp){
        count++;
        temp = temp->next;
    }
    return count;
}

bool searchItem(Node* head, int item){
    Node* temp = head;
    while(temp){
        if(temp->data == item) return true;
        temp = temp->next;
    }
    return false;
}
int main() {
    cout << "Learning linked list"<<endl;
    vector<int> arr = {1,2,3,4,5};
    traverseLL(convertArrToLL(arr));
    int length = countLength(convertArrToLL(arr));
    cout<<"The length of LL is "<<length<<endl;
    cout<<"The item "<<3<<"present is in LL "<<searchItem(convertArrToLL(arr),9)<<endl;
    return 0;
}

File: Linked List/linked.ts
class ListNode{
    data: number;
    next: ListNode | null;
    constructor(data: number, next: ListNode | null = null){
        this.data = data;
        this.next = next;
    }
}

let node1 = new ListNode(2);
let node2 = new ListNode(3,node1);
// console.log(node1)
// console.log(node2)

const arr = [1,0];
// const arr = [1,9,3,4,5,0];
const convertToLL = (arr: number[]) : ListNode => {
    let head = new ListNode(arr[0]);
    let mover = head;
    for(let i = 1; i < arr.length; i++){
        let temp = new ListNode(arr[i]);
        mover.next = temp
        mover = temp
    }
    return head;
}

const traverseLL = (head : ListNode) : void => {
    let temp : ListNode | null = head;
    while(temp){
        console.log(temp.data);
        temp = temp.next;
    }
}

traverseLL(convertToLL(arr))

const countLengthOfLL = (head : ListNode) : number => {
    let count = 0;
    let temp: ListNode | null = head;
    while(temp){
        count++;
        temp = temp.next
    }
    return count;
}
const lengthOfLL = countLengthOfLL(convertToLL(arr));
console.log(lengthOfLL)

const searchItem = (head: ListNode, item: number) : Boolean => {
    let temp: ListNode | null = head;
    while(temp){
        if(temp.data === item) return true;
        temp = temp.next;
    }
    return false;
}
console.log(searchItem(convertToLL(arr),9))

const deleteFirstItem = (head: ListNode) : ListNode | null => {
    if(!head) return null;
    return head.next;
}

console.log(deleteFirstItem(convertToLL(arr)))